core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+        self.state = []
+        self.host = config.get("BLOCKCHAIN", "host")
+        self.ipfs_port = config.get("BLOCKCHAIN", "ipfs_port")
+        self.port = config.get("BLOCKCHAIN", "http_port")
+        self.client = None
+        try:
+            self.client = ipfsapi.connect(self.host, self.ipfs_port)
+        except Exception as e:
+            logging.info("IPFS daemon not started, got: {0}".format(e))
+
+        self.CALLBACKS = {
+            MessageEventTypes.NEW_WEIGHTS.name: self.broadcast_new_weights, 
+            MessageEventTypes.NOTHING.name: self._do_nothing,
+        }
+
+    ##########################################################################
+    ###                            API SECTION                             ###
+    ##########################################################################
+
+    def _do_nothing(self, payload: dict):
+        """
+        Do nothing.
+        """
+        pass
+
+    def get_global_state(self):
+        """
+        Gets the global state which should be a list of dictionaries
+        """
+        timeout = time.time() + 5
+        tx_receipt = None
+        while time.time() < timeout:
+            try:
+                tx_receipt = requests.get(
+                    "http://localhost:{0}/state".format(self.port))
+                tx_receipt.raise_for_status()
+                if tx_receipt:
+                    break
+            except (UnboundLocalError, requests.exceptions.ConnectionError) as e:
+                logging.info("HTTP GET error, got: {0}".format(e))
+                continue
+        logging.info("global state:{}".format(tx_receipt.json()))
+        return tx_receipt.json()
+
+    def get_diffs(self, old_state: list, new_state: dict) -> list:
+        """
+        Iterate through oldState and newState to see any differences
+        Take action based on the differences
+        """
+        tx_diffs = [tx for tx in new_state.get('messages') if (
+                        tx not in old_state)]
+        return tx_diffs
+
+    def get_state_diffs(self, event_filter: object) -> list:
+        """
+        Gets state, then finds diffs, then sets state of blockchain.
+        """
+        new_state = self.get_global_state()
+        if new_state:
+            state_diffs = self.get_diffs(self.state, new_state)
+            filtered_diffs = [tx for tx in state_diffs if event_filter(tx)]
+            logging.info("filtered diffs:{}".format(filtered_diffs))
+            self.update_state(new_state)
+            return filtered_diffs
+
+    def check_malformed_content(self) -> None:
+        """
+        Checks whether content is of the form
+            `{
+                type: ... ,
+                payload: ...
+            }`
+        """
+        pass
+
+    def update_state(self, new_state_wrapper: list) -> None:
+        """
+        Given the freshly-downloaded state, call a handler on each transaction
```

Comment:
The docs should explain or at least mention what the `new_state_wrapper` parameter is.

Reply:
@nzoghb i have no clue what new_state_wrapper is

Reply:
Lotion's state has to come in a superficial dict. Generally state can have other keys besides `messages` but we don't need anything additional

Reply:
@georgymh lmk if you can think of a better name